src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/FaultDetectionMetricsProcessor.java

```diff
@@ -0,0 +1,142 @@
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.reader_writer_shared.Event;
+import java.io.File;
+import java.sql.Connection;
+import java.util.Map;
+import java.util.NavigableMap;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jooq.BatchBindStep;
+
+public class FaultDetectionMetricsProcessor implements EventProcessor {
+    private static final Logger LOG = LogManager.getLogger(FaultDetectionMetricsProcessor.class);
+    private FaultDetectionMetricsSnapshot faultDetectionMetricsSnapshot;
+    private long startTime;
+    private long endTime;
+    private BatchBindStep handle;
+
+    public FaultDetectionMetricsProcessor(FaultDetectionMetricsSnapshot faultDetectionMetricsSnapshot) {
+        this.faultDetectionMetricsSnapshot = faultDetectionMetricsSnapshot;
+    }
+
+    static FaultDetectionMetricsProcessor buildFaultDetectionMetricsProcessor(
+            long currWindowStartTime,
+            Connection conn,
+            NavigableMap<Long, FaultDetectionMetricsSnapshot> faultDetectionMetricsMap) {
+        if (faultDetectionMetricsMap.get(currWindowStartTime) == null) {
+            FaultDetectionMetricsSnapshot faultDetectionMetricsSnapshot =
+                    new FaultDetectionMetricsSnapshot(conn, currWindowStartTime);
+            Map.Entry<Long, FaultDetectionMetricsSnapshot> entry = faultDetectionMetricsMap.lastEntry();
+            if (entry != null) {
+                faultDetectionMetricsSnapshot.rolloverInFlightRequests(entry.getValue());
+            }
+            faultDetectionMetricsMap.put(currWindowStartTime, faultDetectionMetricsSnapshot);
+            return new FaultDetectionMetricsProcessor(faultDetectionMetricsSnapshot);
+        } else {
+            return new FaultDetectionMetricsProcessor(faultDetectionMetricsMap.get(currWindowStartTime));
+        }
+    }
+
+    @Override
+    public void initializeProcessing(long startTime, long endTime) {
+        this.startTime = startTime;
+        this.endTime = endTime;
+        this.handle = faultDetectionMetricsSnapshot.startBatchPut();
+    }
+
+    @Override
+    public void finalizeProcessing() {
+        if (handle.size() > 0) {
+            handle.execute();
+        }
+        LOG.debug("Final Fault Detection request metrics {}", faultDetectionMetricsSnapshot.fetchAll());
+    }
+
+    @Override
+    public void processEvent(Event event) {
+        String[] keyItems = event.key.split(File.separatorChar == '\\' ? "\\\\" : File.separator);
+        if (keyItems[0].equals(PerformanceAnalyzerMetrics.sFaultDetection)) {
+            if (keyItems[3].equals(PerformanceAnalyzerMetrics.START_FILE_NAME)) {
+                emitStartMetric(event, keyItems);
+            } else if (keyItems[3].equals(PerformanceAnalyzerMetrics.FINISH_FILE_NAME)) {
```

Comment:
I assume the `keyItems` or `event.key` will always have the expected values an index 0 and 3 here ? 
What happens if `keyItems[3]` throws an `ArrayIndexOutOfBounds` Exception?

Reply:
That shouldn't be the case. If this happens somehow, it will have an ArrayIndexOutOfBoundsException. Can add assert here which can check the size of keyItems.